awx/main/access.py

```diff
@@ -1923,19 +1923,91 @@ def can_change(self, obj, data):
         return self.check_related('inventory', Inventory, data, obj=obj, role_field='use_role')
 
     def can_attach(self, obj, sub_obj, relationship, data, skip_sub_obj_read_check=False):
+        try:
+            obj_name = obj.name
+        except AttributeError:
+            obj_name = obj.identifier
+
         if isinstance(sub_obj, Credential) and relationship == 'credentials':
-            return self.user in sub_obj.use_role
-        else:
-            raise NotImplementedError('Only credentials can be attached to launch configurations.')
+            if not self.user in sub_obj.use_role:
+                logger.debug(
+                    "User {} not allowed access to credential {} for {} {} ({})".format(self.user.username, sub_obj.name, obj.__class__, obj_name, obj.id)
+                )
+                return False
+            return True
+
+        if isinstance(sub_obj, Label) and relationship == 'labels':
+            if not self.user.can_access(Label, 'read', sub_obj):
+                logger.debug("User {} not allowed access to label {} for {} {} ({})".format(self.user.username, sub_obj.name, obj.__class__, obj_name, obj.id))
+                return False
+            return True
+
+        if isinstance(sub_obj, InstanceGroup) and relationship == 'instance_groups':
+            if not sub_obj in self.user.get_queryset(InstanceGroup):
+                logger.debug(
+                    "User {} not allowed access to instance_group {} for {} {} ({})".format(self.user.username, sub_obj.name, obj.__class__, obj_name, obj.id)
+                )
+                return False
+            return True
```

Comment:
I get the need for some divergences from the parent class here:

https://github.com/ansible/awx/blob/9203062cd64b929cbc6d692066aec7d2e58ab63f/awx/main/access.py#L284-L288

But I'd like to sketch out the goals here from the first principles. In this change, you added `instance_groups` to WFJT nodes, not just schedules. This is evidenced by the introduction of the view `WorkflowJobTemplateNodeInstanceGroupsList`, likewise with schedules.

But WFJT node access class makes no reference to the access method here. So why bother? Or is it broken?

There's also a high degree of repetition here. If the node identifier is a big problem, then fix it in `BaseAccess.can_attach`, and it's really hard for me to see the need for the subclass to handle much... other than not checking the primary object can_change (because it's a supplemental check), but this makes it hard to justify keeping this method at all, since the `BaseAccess` implementation will probably be fine as it is manifested for `ScheduleAccess.can_attach` and the like.